### PR TITLE
Cleanup old `addons` `ManagedResource`

### PR DESCRIPTION
**Pull request by shafeeqes**
**How to categorize this PR?**

/area usability quality
/kind cleanup

**What this PR does / why we need it**:
The addons were refactored with https://github.com/gardener/gardener/pull/7267 and https://github.com/gardener/gardener/pull/7294 and both got released with `v1.63`. The MR `addons` was dropped with https://github.com/gardener/gardener/pull/7574 and released with `v1.66`.

But if the Shoot was hibernated in between `v1.63` and `v1.66`, then the migration will not be complete. https://github.com/gardener/gardener/blob/49dc216c1f6ee36d7893a1eecad407e60e4ed30e/pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go#L554
And if the Shoot is woken up after `v1.66`, It just tries to delete the MR containing the old resources which could interfere with the new MRs `shoot-addon-nginx-ingress` and `shoot-addon-kubernetes-dashboard`.

Furthermore, the cleanup code was removed with https://github.com/gardener/gardener/pull/7868 and `v1.71.0`
So if a shoot was hibernated from `v1.66.0` (or before) all the way through `v1.71.0`, the addon MR can still be present.

This PR first tries to migrate the MR and then if the migration is complete, the MR will be deleted in the next reconciliation.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
We need to keep this for more releases, since the same issue can happen for hibernated Shoots with this PR too.
/cc @acumino @rfranzke @ary1992 

**Release note**:

```bugfix operator
The obsolete `addons` `ManagedResource` is now properly cleaned up.
```

### CODE REVIEW

**Comment by rfranzke** [comment]
> We need to keep this for more releases, since the same issue can happen for hibernated Shoots with this PR too.

How do we prevent that we run again into such situations? We cannot know/assume that all shoots have been woken up at least once before cleaning it up again.
Why didn't you follow the suggestion to do it in the start-up code of gardenlet?

**Comment by shafeeqes** [comment]
> > We need to keep this for more releases, since the same issue can happen for hibernated Shoots with this PR too.
> 
> How do we prevent that we run again into such situations? We cannot know/assume that all shoots have been woken up at least once before cleaning it up again. Why didn't you follow the suggestion to do it in the start-up code of gardenlet?

Even if we do it in the startup, there is no `gardener-resource-manager` pod running for a hibernated Shoot.
The Shoot has to be woken up for the MR to be deleted.

**Comment by rfranzke** [comment]
OK, perhaps it's much easier to simply "force-delete" these resources in the start-up (delete + remove finalizer)? We just want to clean it up and we don't want GRM to really do anything, right? This way
- we know it will work within one release even if the shoot is still hibernated
- we don't need all this code that recreates resources with "ignore" annotation

**Comment by shafeeqes** [comment]
Good idea, but we can do this also only in the shoot reconciliation and not gardenlet startup.
Because there could be some Shoots which are awake in which the `addons` MR is present and GRM is running (because the shoot was woken up after `v1.71`). If we delete the MR, it will try to clean up the resources and some resources can get stuck in terminating, like the `kubernetes-dashboard` namespace.

Or else we'll have to check the Shoot if it's hibernated in the gardenlet startup. Either way two handlings are needed.

**Comment by rfranzke** [comment]
This has the downside that shoots stuck in failed (because of invalid infrastructure credentials) are not covered unless you do it before the shoot flow starts, but then you can also do it in the startup code.

To make it as easy as possible, what about implementing the following:
- For all shoot namespaces where this `addons` MR still exists, scale down GRM to `0`, force-delete the MR, scale-up GRM again (if it was >0 before)

File: cmd/gardenlet/app/app.go [hunk 1]
```diff
@@ -557,6 +568,83 @@ func (g *garden) cleanupStaleShootStates(ctx context.Context, gardenClient clien
...
+				if err := seedClient.Patch(ctx, resourceManagerDeployment, patch); err != nil {
+					return fmt.Errorf("failed to scale gardener-resource-manager deployment to zero %q: %w", client.ObjectKeyFromObject(resourceManagerDeployment), err)
+				}
+			}
+
+			patch := client.MergeFrom(addonsMR.DeepCopy())
+			addonsMR.Finalizers = nil
+			if err := seedClient.Patch(ctx, addonsMR, patch); err != nil {
+				return fmt.Errorf("failed to patch ManagedResource %q: %w", client.ObjectKeyFromObject(addonsMR), err)
+			}
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: Nit: You could use https://github.com/gardener/gardener/blob/5abed1bfedd6f2188e5dcec9af161d1b5c286236/pkg/controllerutils/finalizers.go#L40-L46

File: cmd/gardenlet/app/app.go [hunk 2]
```diff
@@ -557,6 +568,83 @@ func (g *garden) cleanupStaleShootStates(ctx context.Context, gardenClient clien
...
+
+			if err := managedresources.DeleteForShoot(ctx, seedClient, namespace.Name, "addons"); err != nil {
+				return err
+			}
+
+			if err := managedresources.WaitUntilDeleted(ctx, seedClient, namespace.Name, "addons"); err != nil {
+				return err
+			}
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: This waits forever. Also, why is it needed in the first place? We have removed all finalizers, so above delete call will make the resources to immediately disappear from etcd?

File: cmd/gardenlet/app/app.go (hunk 1 above)
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: Don't we need to do the same for the resp. secret?

Comment (shafeeqes) [reply cmd/gardenlet/app/app.go]: We delete the secret and GRM will remove finalizer if the secret is not referenced by any MRs.
But considering that the shoot could be hibernated for a long time, we can also finalize this secret right away.

File: cmd/gardenlet/app/app.go [hunk 3]
```diff
@@ -557,6 +568,93 @@ func (g *garden) cleanupStaleShootStates(ctx context.Context, gardenClient clien
...
+
+			if err := seedClient.Get(ctx, client.ObjectKeyFromObject(addonsMR), addonsMR); err != nil {
+				// If the MR is already gone, then nothing to do here
+				if apierrors.IsNotFound(err) {
+					return nil
+				}
+				return err
+			}
+
+			if err := seedClient.Get(ctx, client.ObjectKeyFromObject(resourceManagerDeployment), resourceManagerDeployment); err != nil {
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: Just to be on the safe side:
```suggestion
			if err := seedClient.Get(ctx, client.ObjectKeyFromObject(resourceManagerDeployment), resourceManagerDeployment); err != nil {
				if apierrors.IsNotFound(err) {
					return nil
				}
```

File: cmd/gardenlet/app/app.go [hunk 4]
```diff
@@ -557,6 +568,93 @@ func (g *garden) cleanupStaleShootStates(ctx context.Context, gardenClient clien
...
+			if err := controllerutils.RemoveAllFinalizers(ctx, seedClient, addonsMR); err != nil {
+				return fmt.Errorf("failed to remove finalizers from the ManagedResource %q: %w", client.ObjectKeyFromObject(addonsMR), err)
+			}
+
+			if err := seedClient.Get(ctx, client.ObjectKeyFromObject(addonsSecret), addonsSecret); err == nil {
+				if err := controllerutils.RemoveAllFinalizers(ctx, seedClient, addonsSecret); err != nil {
+					return fmt.Errorf("failed to remove finalizers from the Secret %q: %w", client.ObjectKeyFromObject(addonsSecret), err)
+				}
+			} else {
+				if !apierrors.IsNotFound(err) {
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]:
```suggestion
			} else if !apierrors.IsNotFound(err) {
```

**Pull request merged**